fv3config/config/diag_table.py

```diff
@@ -0,0 +1,212 @@
+import logging
+from typing import Sequence, Optional, Union
+import dataclasses
+import datetime
+import re
+
+import dacite
+
+from .._exceptions import ConfigError
+
+logger = logging.getLogger("fv3config")
+
+
+@dataclasses.dataclass
+class DiagTableField:
+    """Object representing a field of a diagnostics file."""
+
+    module_name: str
+    field_name: str
+    output_name: str
+    reduction_method: Optional[str] = "none"
+    time_sampling: Optional[str] = "all"
+    regional_section: Optional[str] = "none"
+    packing: Optional[int] = 2
+
+
+@dataclasses.dataclass
+class DiagTableFile:
+    """Object representing a diagnostics file."""
+
+    name: str
+    frequency: int
+    frequency_units: str
```

Comment:
Because time frequency can only be one of "years", "months", "days", "minutes", "hours", or "seconds", I would suggest using an [enum class](https://docs.python.org/3/library/enum.html) for this. This could also be used for the time_axis, if they have the same possibilities. I believe this is [supported by dacite](https://github.com/konradhalas/dacite/pull/65).

Reply:
Or even better use a datetime.timedelta object and define a custom parser for frequencies:

```
>>> import dacite
... import dataclasses
... import datetime
...
...
... def parse_str(duration):
...     return datetime.timedelta(hours=3)
...
...
... @dataclasses.dataclass
... class A:
...     freq: datetime.timedelta
...
... a = dacite.from_dict(A, {"freq": "3h"}, config=dacite.Config(type_hooks={datetime.timedelta: parse_str}))
>>> a
A(freq=datetime.timedelta(seconds=10800))
```

Reply:
Certain frequency int values such as 0 or -1 have special meanings for the diag_table, which is why I kept the `frequency` and `frequency_units` attributes separate.

Reply:
Couldn't you identify those 0 with `0hr` and -1 with any negative value of `timedelta`?

Reply:
We could... but to me it just introduces more levels of documentation required, e.g. specifying how we expect the timedelta to be serialized ("3h") whereas right now it's clear.

Also the Fortran model has a concept of a "month" time frequency/interval which is not covered by python datetime.timedelta.

Reply:
Put another way, the docs available for the diag_table are actually quite good (e.g. https://mom6.readthedocs.io/en/dev-gfdl/api/generated/pages/Diagnostics.html) and so I'm trying to stick to the established spec as much as possible, but with a more hierarchal structure and with sensible defaults. That way we don't have to redocument or learn a new system.

Reply:
I love replacing Fortran "time amount plus units" with timedeltas. If we're not refactoring the "special meanings" into a different attribute representation (like boolean flags) and want to stick to the existing docs, I'm supportive of keeping this as-is.

If we want a timedelta, it may be worthwhile adding a property method which does the unit string parsing and provides a timedelta (but raises an exception for months).